### ISSUE TITLE
Buildstream local cache doesn't get cleanup automatically when the disk is full

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#737]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/737)
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Oct 29, 2018, 15:38

## Summary
If the disk gets fill up, build will fail and the only solution is to `rm -r` the buildstream cache folder

## What is the expected correct behavior?
If the disk is full, buildstream should automatically cleanup, similar to what what happen when you set a quota

## Possible fixes
A possible solution would be (when we fix #733) to change the defaults from
```
quota: infinity
```
to
```
quota: 100%
```
* BuildStream version affected: 1.2.3

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @toscalix]](https://gitlab.com/toscalix) on Nov 1, 2018, 09:29

marked this issue as related to #700

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jennis]](https://gitlab.com/jennis) on Nov 7, 2018, 16:41

marked this issue as related to #754

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jennis]](https://gitlab.com/jennis) on Nov 7, 2018, 16:43

marked this issue as related to #733

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tlater]](https://gitlab.com/tlater) on Nov 8, 2018, 13:01

Would you mind sharing logs for this? BuildStream *should* already do this.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jan 19, 2019, 16:41

BuildStream is trusting it's artifact cache to remember it's size (the `cache_size` file), which is fine, but BuildStream does not know if external conditions have changed, or how many torrents you had downloading in parallel in that adjacent directory.

I think BuildStream also does not force the check at startup, instead a build is launched right away and the cache size job / cleanup comes after the build completes.

We could start by:
* Checking if our saved cache_size exceeds the quota
* If the cache size exceeds the quota, launch an *exclusive* cache size job initially for any session which involves growing the cache
* Then launch the session

If the cache size needs to be calculated, it will happen before anything else beats it to the CACHE scheduling resource.

Asides from this, it might make sense to add a new kind of cache size job which only checks the status of `df` for the cache partition, to monitor if enough space still remains every so often as the build progresses...

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jan 20, 2019, 00:34

mentioned in commit 3dd8fb6714d9fdef8fcf59d043fbe82bbcacfea2

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jan 20, 2019, 20:03

mentioned in commit edb4933eabe954c99f53cf6a72d7ab8e10bf3166

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jan 20, 2019, 20:05

mentioned in commit 9493a0f54a94349ac3be78e1e2bbc7c7422deb37

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jan 22, 2019, 18:23

mentioned in commit 7a3f6cbcbc194afc3d08c20c1103d48d9a05e04d

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jan 22, 2019, 20:00

mentioned in commit 3f4a68d54f06993e17ef2d24ebcbf1d20885556b

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jan 23, 2019, 13:58

mentioned in commit 441e8dd0c94ee973fbcabadb82c57b7e5aded35a

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jan 24, 2019, 16:58

mentioned in commit ce01f87e3ac759b12eece4b20be1987ac0e6151c

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Feb 27, 2019, 15:12

I believe this is fixed by !1106.

Since !1106, the `ArtifactCache` reports that it is *full* instead of whether it's quota has been exceeded, and the scheduler uses this both at startup to ensure we start with a cleanup if needed, or at any time we calculate the cache size.

At all times, we consider a condition where the remaining disk space is less than the hard coded headroom of `2G` as a condition where a cleanup is needed.

* If we want to make the headroom configurable, that should be a separate issue.

* If we are seeing stack traces due to exceeding disk space *anyways* even though we have done our best to clean up in time, then we are in fact talking about #660, and need not duplicate that issue here.

[[Gitlab user @jjardon]](https://gitlab.com/jjardon) Are you satisfied with closing this issue ?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Feb 28, 2019, 09:11

[[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) sure, thanks!

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Feb 28, 2019, 09:11

closed
